lib/CL/pocl_binary.c

```diff
@@ -0,0 +1,615 @@
+      }                                           \
+  } while (0)
+
+#define BUFFER_READ_STR2(elem, len)               \
+  do {                                            \
+    BUFFER_READ(len, uint32_t);                   \
+    if (len)                                      \
+      {                                           \
+        elem = malloc(len+1);                     \
+        memcpy(elem, buffer, len);                \
+        elem[len] = 0;                            \
+        buffer += len;                            \
+      }                                           \
+  } while (0)
+
+#define BUFFER_STORE_STR(elem)                    \
+  do { uint32_t len = strlen(elem);               \
+    BUFFER_STORE_STR2(elem, len); } while (0)
+
+#define BUFFER_READ_STR(elem)                     \
+  do { uint32_t len = 0;                          \
+    BUFFER_READ_STR2(elem, len); } while (0)
+
+#define ADD_STRLEN(else_b)                        \
+    if (serialized)                               \
+      {                                           \
+        unsigned char* t = buffer+res;            \
+        res += *(uint32_t*)t;                     \
+        res += sizeof(uint32_t);                  \
+      }                                           \
+    else                                          \
+      {                                           \
+        res += sizeof(uint32_t);                  \
+        res += else_b;                            \
+      }
+
+/***********************************************************/
+
+static unsigned char* read_header(pocl_binary *b, const unsigned char *buffer)
+{
+  memset(b, 0, sizeof(pocl_binary));
+  BUFFER_READ(b->endian, char);
+  memcpy(b->pocl_id, buffer, POCLCC_STRING_ID_LENGTH);
+  buffer += POCLCC_STRING_ID_LENGTH;
+  BUFFER_READ(b->device_id, uint64_t);
+  BUFFER_READ(b->version, uint32_t);
+  BUFFER_READ(b->num_kernels, uint32_t);
+  memcpy(b->program_build_hash, buffer, sizeof(SHA1_digest_t));
+  buffer += sizeof(SHA1_digest_t);
+  return (unsigned char*)buffer;
+}
+
+#define FNV_OFFSET UINT64_C(0xcbf29ce484222325)
+#define FNV_PRIME UINT64_C(0x100000001b3)
+static uint64_t pocl_binary_get_device_id(cl_device_id device)
+{
+  //FNV-1A with vendor_id, llvm_target_triplet and llvm_cpu
+  uint64_t result = FNV_OFFSET;
+  const char *llvm_tt = device->llvm_target_triplet;
+  const char *llvm_cpu = device->llvm_cpu;
+
+  result *= FNV_PRIME;
+  result ^= device->vendor_id;
+  int i, length = strlen(llvm_tt);
+  for (i=0; i<length; i++)
+    {
+      result *= FNV_PRIME;
+      result ^= llvm_tt[i];
+    }
+  length = strlen(llvm_cpu);
+  for (i=0; i<length; i++)
+    {
+      result *= FNV_PRIME;
+      result ^= llvm_cpu[i];
+    }
+
+  return result;
+}
+
+static unsigned char* check_binary(cl_device_id device, const unsigned char *binary)
+{
+  pocl_binary b;
+  unsigned char *p = read_header(&b, binary);
+  if (b.endian != host_endian)
```

Comment:
So you require endianness? The binary should be of a single endianness instead. Or is there a reason why not? Currently probably implementation burden reasons only as you need to swap all the fields conditionally?